### PR TITLE
Adds support for setting user labels on functions

### PR DESCRIPTION
**Pull request by joehan**
### Description
Adds 'labels' to runtimeOptions, per go/firebase-functions-labels. This lets you set user labels on functions!

Note: I did make one minor change to the behavior described in go/firebase-functions-labels - if the labels option is not set in runWith(), we still overwrite labels. Originally, I proposed this behavior to support users who already manually set labels. However, the existing behavior of Firebase Functions is to overwrite labels. We also avoid the awkward case where removing 'labels' from 'runtimeOptions' doesn't cause labels to be removed. 

### Code sample
```
functions.runWith({
  labels: {
    "my-label": "my-value",
  }
})...
```

### Testing
I tested the following cases and confirmed that labels were set correctly via the Equivalent REST button in GCP console.
- Setting labels on normal functions.
- Setting labels on scheduled functions.
- Updating a function to change the labels.
- Setting too many labels throws an error before deployment
- Setting invalid labels throws an error before deployment

### CODE REVIEW

File: src/function-builder.ts [hunk 1]
```diff
@@ -120,6 +121,61 @@ function assertRuntimeOptionsValid(runtimeOptions: RuntimeOptions): boolean {
...
+    );
+    if (invalidLengthValues.length > 0) {
+      throw new Error(
+        `Invalid labels: ${invalidLengthValues.join(
+          ', '
+        )}. Label values must be less than 64 charcters.`
+      );
+    }
+
+    // Keys can contain lowercase letters, foreign characters, numbers, _ or -. They must start with a letter.
```
Comment (taeold) [inline src/function-builder.ts:154]: nit: `They must start with a lowercase letter`

Also, I'd appreciate if we link https://cloud.google.com/resource-manager/docs/creating-managing-labels#requirements somewhere!

File: src/function-configuration.ts [hunk 2]
```diff
@@ -147,6 +149,11 @@ export interface RuntimeOptions {
    * Ingress settings
    */
   ingressSettings?: typeof INGRESS_SETTINGS_OPTIONS[number];
+
+  /**
+   *
```
Comment (taeold) [inline src/function-configuration.ts:154]: missing documentation?

File: src/function-builder.ts [hunk 3]
```diff
@@ -120,6 +121,61 @@ function assertRuntimeOptionsValid(runtimeOptions: RuntimeOptions): boolean {
       `serviceAccount must be set to 'default', a service account email, or '{serviceAccountName}@'`
     );
   }
+
```
Comment (taeold) [inline src/function-builder.ts:124]: Should we explicitly disallow implicitly reserved labels? e.g. I think we reserve the use of `deployment-tools` (as in, we are going to overwrite whatever you put as `deployment-tools` with whatever the CLI wants it to be.

File: src/cloud-functions.ts [hunk 4]
```diff
@@ -417,7 +417,7 @@ export function makeCloudFunction<EventData>({
         },
       });
       if (!_.isEmpty(labels)) {
-        trigger.labels = labels;
+        trigger.labels = { ...labels, ...trigger.labels };
```
Comment (inlined) [inline src/cloud-functions.ts:420]: I think I'd actually flip this. I'd rather bugs show up in the form of "Why can't I set label X" as opposed to "Why am I seeing arbitrary mysterious behavior that depended on a reserved label"

File: src/function-builder.ts (hunk 3 above)
Comment (inlined) [reply src/function-builder.ts:124]: I like this. I'd also possibly disable any label that starts with `firebase-`

File: src/function-configuration.ts [hunk 5]
```diff
@@ -94,6 +94,8 @@ export const DEFAULT_FAILURE_POLICY: FailurePolicy = {
   retry: {},
 };
 
+export const MAX_NUMBER_USER_LABELS = 58;
```
Comment (inlined) [inline src/function-configuration.ts:97]: Does this actually need to be exported here (and thus in index.ts)? We could just put this const in cloudfunctions.ts

**Review by inlined — APPROVED** [review]

File: src/function-builder.ts (hunk 1 above)
Comment (joehan) [reply src/function-builder.ts:154]: Good call, I'll add that link. However, they can start with an international character as well -technically, anything in the Unicode groups 'letters lowercase' or 'letters other'

File: src/function-configuration.ts (hunk 5 above)
Comment (joehan) [reply src/function-configuration.ts:97]: We don't technically need to. However, we export other similar constants here, like MAX_TIMEOUT_SECONDS, so I figured it was appropriate to do so for this one too

File: src/function-configuration.ts (hunk 2 above)
Comment (joehan) [reply src/function-configuration.ts:154]: Good catch. I also went through and cleaned up some of these other, less descriptive comments.

**Pull request merged**